### ISSUE TITLE
If training set has same name and tags in different directories wrong tags are used.

### ISSUE THREAD

**Issue by v-vincere** [ultralytics/yolov5#5259]
As the title says wrong bounding box text files are being used when they share the same name even if they are in different directories.

Example:
I have a folder that is called dataset and within this folder is three folders that contains,

```
dataset-
   -1
      -frame1.txt
      -frame1.png
      -frame2.txt
      -frame2.png
      -frame3.txt
      -frame3.png
   -2
      -frame1.txt
      -frame1.png
      -frame2.txt
      -frame2.png
      -frame3.txt
      -frame3.png
   -3
      -frame1.txt
      -frame1.png
      -frame2.txt
      -frame2.png
      -frame3.txt
      -frame3.png
```

when I add folders 1, 2, 3 into my training configs.yaml and run training with cache enabled my training batch mosaics will have wrong bounding boxes because it is taking the frame1.png (from folder 1) and using the bounding box text file frame1.txt (from folder 3 instead of folder 1).

This obviously results in poor training. My workaround that fixes this issue is running a script that changes the filename in each folder so no file across my dataset shares the same name. What are your thoughts on fixing this issue without needed to run a script that renames files?

**Comment by glenn-jocher** [comment]
@idkvince hi, thank you for your feature suggestion on how to improve YOLOv5 🚀! This is a good question. It's recommended for your images to all have unique names within a dataset.

If you'd like to implement an improvement though, the fastest and easiest way to incorporate your ideas into the official codebase is to submit a Pull Request (PR) implementing your idea, and if applicable providing before and after profiling/inference/training results to help us understand the improvement your feature provides. This allows us to directly see the changes in the code and to understand how they affect workflows and performance.

Please see our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/) to get started.

**Comment by v-vincere** [comment]
Hi @glenn-jocher thanks for the update, I'm afraid the script I have that modifies the file name isn't exactly robust and will not pass your code quality checks. I just wanted to bring this issue to your attention and help others who might run into this problem in the future.

**Comment by glenn-jocher** [comment]
@idkvince thanks! Do you know exactly where this is failing in the caching/dataloader? 

Maybe I can do the PR myself, as I'm sure you are not the only one who may be affected. The dataloader uses full image paths for each image, so on the surface it looks like it should be capable of handling the above scenario. 

<img width="1521" alt="Screenshot 2021-10-22 at 18 11 07" src="https://user-images.githubusercontent.com/26833433/138488143-4fb1dbba-922f-4939-9fd3-fe151172196b.png">

**Comment by glenn-jocher** [comment]
@idkvince full labelfile paths are also used, i.e. `self.img_labels`:
<img width="1521" alt="Screenshot 2021-10-22 at 18 11 47" src="https://user-images.githubusercontent.com/26833433/138488255-d2d45fcb-80df-4a51-aea9-3b48fe2ae9dd.png">

**Comment by glenn-jocher** [comment]
@idkvince I attempted to reproduce and am not able to. Your above setup works correctly for me. Annotations are correct even for same-filename images. It seems that if there is a problem it is with your labels and not YOLOv5.

<img width="1218" alt="Screenshot 2021-10-22 at 18 41 41" src="https://user-images.githubusercontent.com/26833433/138492107-d5050bb3-67f6-4a16-8324-6db7fe3b1262.png">

```yaml
# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: ../datasets/coco128  # dataset root dir
train: [a, b]  # train images (relative to 'path') 128 images
val: a  # val images (relative to 'path') 128 images
test:  # test images (optional)
```

![image](https://user-images.githubusercontent.com/26833433/138491966-305dd3cb-ff8b-458e-8d6a-c607181df696.png)

**Comment by v-vincere** [comment]
Hi @glenn-jocher I was just able to repeat the errant tags in my data set. My data is over 100k images across 15 directories, it only happens when I run '--cache' without cache the bad tags don't appear. 

My steps that I did was I reverted my names so there were duplicate names across directories and trained with --cache and saw the bad tags in train_batch*.jpg. I then deleted the cache files and ran training again without --cache and saw that the tags were fine.

**Comment by glenn-jocher** [comment]
@idkvince yes in my above example I create the dataset as shown with no *.cache file and then tried to reproduce with these two commands, deleting the *.cache file each time, but everything still worked in both cases.
```bash
python train.py
python train.py --cache
```

If it's not working for you, and you are using unmodified current code (not out of date), then please try to submit a reproducible example so that we can reproduce on our side. If we can't reproduce there's not much for us to do.

We've created a few short guidelines below to help users provide what we need in order to get started investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

* ✅ **Minimal** – Use as little code as possible that still produces the same problem
* ✅ **Complete** – Provide **all** parts someone else needs to reproduce your problem in the question itself
* ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

In addition to the above requirements, for [Ultralytics](https://ultralytics.com/) to provide assistance your code should be:

* ✅ **Current** – Verify that your code is up-to-date with current GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been resolved by previous commits.
* ✅ **Unmodified** – Your problem must be reproducible without any modifications to the codebase in this repository. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all of the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) and providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃
